src/rewritable_units/mutations.rs

```diff
@@ -12,83 +16,169 @@ pub enum ContentType {
     }
 
     #[inline]
-    pub fn before(&mut self, content: &str, content_type: ContentType) {
-        content_to_bytes(content, content_type, self.encoding, &mut |c| {
-            self.content_before.extend_from_slice(c);
-        });
+    pub fn removed(&self) -> bool {
+        self.inner.as_ref().is_some_and(|inner| inner.removed)
+    }
+}
+
+impl From<(&str, ContentType)> for StringChunk {
+    #[inline]
+    fn from((content, content_type): (&str, ContentType)) -> Self {
+        Self::Buffer(Box::from(content), content_type)
     }
+}
+
+pub(crate) enum StringChunk {
+    Buffer(Box<str>, ContentType),
+    Stream(Box<dyn StreamingHandler>),
+}
+
+#[derive(Default)]
+pub(crate) struct DynamicString {
+    chunks: Vec<StringChunk>,
+}
 
+impl DynamicString {
     #[inline]
-    pub fn after(&mut self, content: &str, content_type: ContentType) {
-        let mut pos = 0;
+    pub const fn new() -> Self {
+        Self { chunks: vec![] }
+    }
 
-        content_to_bytes(content, content_type, self.encoding, &mut |c| {
-            self.content_after.splice(pos..pos, c.iter().cloned());
+    #[inline]
+    pub fn clear(&mut self) {
+        self.chunks.clear();
+    }
 
-            pos += c.len();
-        });
+    #[inline]
+    pub fn push_front(&mut self, chunk: StringChunk) {
+        self.chunks.insert(0, chunk);
     }
 
     #[inline]
-    pub fn replace(&mut self, content: &str, content_type: ContentType) {
-        let mut replacement = Vec::default();
+    pub fn push_back(&mut self, chunk: StringChunk) {
+        self.chunks.push(chunk);
+    }
+
+    pub fn encode(self, sink: &mut StreamingHandlerSink<'_>) -> BoxResult {
+        for chunk in self.chunks {
+            match chunk {
+                StringChunk::Buffer(content, content_type) => {
+                    sink.write_str(&content, content_type);
+                }
+                StringChunk::Stream(handler) => {
+                    handler.write_all(sink)?;
+                }
+            };
+        }
+        Ok(())
+    }
+}
 
-        content_to_bytes(content, content_type, self.encoding, &mut |c| {
-            replacement.extend_from_slice(c);
-        });
+/// A callback used to write content asynchronously.
+pub trait StreamingHandler: Send {
+    /// This method is called only once, and is expected to write content
+    /// by calling the [`sink.write_str()`](StreamingHandlerSink::write_str) one or more times.
+    ///
+    /// Multiple calls to `sink.write_str()` append more content to the output.
+    ///
+    /// See [`StreamingHandlerSink`].
+    fn write_all(self: Box<Self>, sink: &mut StreamingHandlerSink<'_>) -> BoxResult;
 
-        self.replacement = replacement;
-        self.remove();
+    // Safety: due to lack of Sync, this trait must not have `&self` methods
+}
+
+/// Avoid requring `StreamingHandler` to be `Sync`.
+/// It only has a method taking exclusive ownership, so there's no sharing possible.
+unsafe impl Sync for StringChunk {}
```

Comment:
`StreamingHandler` can be implemented for other types, including things that allow sharing and internal mutability, for instance `Arc<RefCell<Foo>>`, so taking self by moves doesn't ensure exclusivity because `self` can be a `Arc<_>`.

Do we actually need `StringChunk` to be `Sync`?

Reply:
`Arc<T>` isn't `Send` if `T` isn't `Sync`, and `StreamingHandler` required `Send`.

I've just removed the `Sync` impl. It wasn't important.